### PR TITLE
[controller] Enforce min backup version cleanup delay at push start

### PR DESCRIPTION
**Pull request by misyel**
## Problem Statement

When a version rollback marks the ex-current version as ERROR, a subsequent new push can delete that ERROR version immediately at start of push (SOP), bypassing the min backup version cleanup delay (default 1h) that exists to let routers/servers/clients switch off the old current version. This can also push the store beyond its provisioned hardware capacity if a new push arrives before backup versions can be safely cleaned up.

More generally, any code path that triggers `retireOldStoreVersions` at SOP doesn't currently honor the min cleanup delay, creating an inconsistency with `StoreBackupVersionCleanupService` (which does enforce it).

## Solution

Add a capacity guard in `VeniceHelixAdmin.addVersion` that blocks a new push if backup versions are pending deletion but cannot be cleaned up yet due to the min cleanup delay.

Behavior differs by backup strategy:
- **DELETE_ON_NEW_PUSH_START**: hardware provisioned for 2x versions (current + new push). Preserve count = N-1.
- **KEEP_MIN_VERSIONS** (and others): hardware provisioned for N+1 where N = configured versions to keep. Preserve count = N (store-level `numVersionsToPreserve` takes precedence when set).

When blocked, the caller receives a clear `VeniceException` with the store name, cluster name, pending version count, and the min cleanup delay value, and can retry after the delay has elapsed.

Also adds a silent skip in `retireOldStoreVersions` when within the min delay as defense-in-depth for the post-push-completion path; the periodic `StoreBackupVersionCleanupService` handles eventual deletion.

### Code changes

- [ ] Added new code behind **a config**. _(Uses existing `CONTROLLER_BACKUP_VERSION_MIN_CLEANUP_DELAY_MS`, default 1h.)_
- [x] Introduced new **log lines**.
  - [x] Confirmed if logs need to be **rate limited** to avoid excessive logging. _(Per-push-start, not per-iteration — no rate limiting needed.)_

### **Concurrency-Specific Checks**

- [x] Code has **no race conditions** or **thread safety issues**. _(Check runs inside existing store write lock.)_
- [x] Proper **synchronization mechanisms** used. _(Inherits existing `createStoreWriteLockOnly` in `addVersion`.)_
- [x] No **blocking calls** inside critical sections.
- [x] Verified **thread-safe collections** are used.
- [x] Validated proper exception handling in multi-threaded code.

## How was this PR tested?

- [x] New unit tests added. _(`TestBackupVersionCleanupCapacityCheck` — 7 tests covering both strategies, preserve-count verification, boundary conditions, and error message content.)_
- [ ] New integration tests added.
- [x] Modified or extended existing tests. _(None needed — existing `TestVeniceHelixAdmin` and `TestStoreBackupVersionCleanupService` still pass.)_
- [x] Verified backward compatibility. _(No enum changes, no config changes, no wire protocol changes.)_

## Does this PR introduce any user-facing or breaking changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Clearly explain the behavior change and its impact.

**Behavior change**: A new push can now be rejected at SOP with a `VeniceException` if:
1. The store has backup versions that would need to be deleted to make room, AND
2. The current version was promoted within the min cleanup delay window (default 1h).

In practice this will only trigger immediately after a rollback (which promotes the backup version and leaves the ex-current as ERROR). Operators can retry the push after the min cleanup delay has elapsed.

### CODE REVIEW

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java [hunk 1]
```diff
@@ -4517,6 +4520,59 @@ private void deleteRTTopicFromAllFabrics(String topic, String clusterName) {
...
+        backupStrategy == BackupStrategy.DELETE_ON_NEW_PUSH_START
+            ? minNumberOfStoreVersionsToPreserve - 1
+            : minNumberOfStoreVersionsToPreserve);
+    List<Version> versionsToDelete = store.retrieveVersionsToDelete(numVersionToPreserve);
+    if (versionsToDelete.isEmpty()) {
+      return;
+    }
+    long minRetentionThreshold = store.getLatestVersionPromoteToCurrentTimestamp() + minBackupVersionCleanupDelay;
+    if (currentTimeMs <= minRetentionThreshold) {
+      throw new VeniceException(
```
Comment (xunyin8) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:4564]: Can we add an integration test ensure VPJ will fail fast upon encountering this exception (instead of hanging or getting stuck in some infinite retry)?

**Review by xunyin8 — CHANGES_REQUESTED** [review]
LGTM just a minor request for additional test.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java [hunk 2]
```diff
@@ -4539,13 +4595,27 @@ public void retireOldStoreVersions(
...
       List<Version> versionsToDelete = store.retrieveVersionsToDelete(numVersionToPreserve);
       if (versionsToDelete.isEmpty()) {
         return;
       }
 
+      // Skip if within min cleanup delay; StoreBackupVersionCleanupService will pick it up later.
+      long minBackupVersionCleanupDelay = multiClusterConfigs.getBackupVersionMinCleanupDelayMs();
+      long minRetentionThreshold = store.getLatestVersionPromoteToCurrentTimestamp() + minBackupVersionCleanupDelay;
+      if (System.currentTimeMillis() <= minRetentionThreshold) {
+        LOGGER.info(
```
[the remark this answers is not in the document]
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:4610]: The concern about unbounded accumulation is mitigated in our production setup: `StoreBackupVersionCleanupService` (the retention-based cleanup service) is always enabled, so any versions that skip deletion in `retireOldStoreVersions` due to the min-delay window get picked up by the periodic service once the delay elapses.

The skip here is intentional: immediately after a push promotes a new version, routers/servers/clients may not yet have switched off the just-promoted predecessor. Deleting at that moment can race with reads. Waiting the min cleanup delay before deletion is the same safety invariant that `StoreBackupVersionCleanupService` already enforces via `pastMinRetention` — this change just extends the invariant to the `retireOldStoreVersions` path that's called post-completion.

The capacity guard at `addVersion` handles the specific rollback-triggered accumulation case (blocking the push until the window elapses), which is the primary motivation for this change.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 1 above)
Comment (misyel) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:4564]: Added `testPushBlockedAfterRollbackWithinMinCleanupDelay` and verified that the push fails when the exception is thrown

**Review by xunyin8 — APPROVED** [review]
